src/widget/wsearchlineedit.cpp

```diff
@@ -200,103 +238,214 @@ void WSearchLineEdit::resizeEvent(QResizeEvent* e) {
+            << "slotTriggerSearch"
+            << getSearchText();
+#endif // ENABLE_TRACE_LOG
+    DEBUG_ASSERT(isEnabled());
     m_debouncingTimer.stop();
     emit search(getSearchText());
 }
 
-void WSearchLineEdit::showPlaceholder() {
+bool WSearchLineEdit::shouldShowPlaceholder(const QString& text) const {
     DEBUG_ASSERT(isEnabled());
+    return text.isEmpty() && m_state != State::Active;
+}
 
-    // Deactivate text change listener
-    m_state = State::Inactive;
-
-    setText(tr("Search...", "noun"));
-
-    QPalette pal = palette();
-    pal.setColor(foregroundRole(), Qt::lightGray);
-    setPalette(pal);
+void WSearchLineEdit::showPlaceholder() {
+#if ENABLE_TRACE_LOG
+    kLogger.trace()
+            << "showPlaceholder"
+            << getSearchText();
+#endif // ENABLE_TRACE_LOG
+    DEBUG_ASSERT(getSearchText() == kEmptySearch);
+    DEBUG_ASSERT(shouldShowPlaceholder(getSearchText()));
+
+    updateClearButton(kEmptySearch);
 }
 
 void WSearchLineEdit::showSearchText(const QString& text) {
+#if ENABLE_TRACE_LOG
+    kLogger.trace()
+            << "showSearchText"
+            << text;
+#endif // ENABLE_TRACE_LOG
     DEBUG_ASSERT(isEnabled());
 
-    // Reactivate text change listener
-    m_state = State::Active;
-
-    // Update the displayed text without (re-)starting the timer
-    blockSignals(true);
-    if (text.isNull()) {
-        setText(kEmptySearch);
+    if (text.isEmpty()) {
+        setTextBlockSignals(kEmptySearch);
     } else {
-        setText(text);
+        setTextBlockSignals(text);
     }
-    blockSignals(false);
-
+    m_state = State::Active;
     updateClearButton(text);
 
-    QPalette pal = palette();
-    pal.setColor(foregroundRole(), m_foregroundColor);
-    setPalette(pal);
-
     // This gets rid of the blue mac highlight.
     setAttribute(Qt::WA_MacShowFocusRect, false);
 }
 
+void WSearchLineEdit::refreshEditBox() {
+#if ENABLE_TRACE_LOG
+    kLogger.trace()
+            << "refreshEditBox";
+#endif // ENABLE_TRACE_LOG
+    if (isEnabled()) {
+        enableSearch(getSearchText());
+    } else {
+        slotDisableSearch();
+    }
+}
+
 void WSearchLineEdit::updateEditBox(const QString& text) {
-    if (text.isEmpty()) {
+#if ENABLE_TRACE_LOG
+    kLogger.trace()
+            << "updateEditBox"
+            << text;
+#endif // ENABLE_TRACE_LOG
+    // Updating the placeholder or search text implicitly updates
+    // both the text and the clear button
+    if (shouldShowPlaceholder(text)) {
         showPlaceholder();
     } else {
         showSearchText(text);
     }
 }
 
 void WSearchLineEdit::updateClearButton(const QString& text) {
-    if (!text.isEmpty() && (m_state == State::Active)) {
-        m_clearButton->setVisible(true);
-        // make sure the text won't be drawn behind the Clear button icon
-        setStyleSheet(QString("QLineEdit { padding-right: %1px; } ")
-                              .arg(m_innerHeight + m_frameWidth));
-    } else {
+#if ENABLE_TRACE_LOG
```

Comment:
I really have difficulties understanding the diff, also while trying to understand the call hierarchy with Eclipse I get the feeling we could concentrate updateClearButton(), showPlaceHolder(), shouldShowPlaceholder() and maybe others into one function. I might be wrong, though..

Reply:
Thanks for the hint. I have removed redundant methods and invocations.